### ISSUE TITLE
New rule proposal: `SafeCallOnNullableTypeInTest`

### ISSUE THREAD

**Issue by yogurtearl** [detekt/detekt#3295]
## Expected Behavior of the rule

Detect usage of `?.` **in a test** that could lead to the test doing nothing at all. 

In this example, if `myFixture.someData` is null here, the test just passes and does nothing because of the `?.`! 
```kotlin
@Test 
fun myTest(){
    val whevs = myFixture.someData?.somefield
    whevs?.let{
          val result = doSomething(whevs) 

          assertStuff(result)
    }
}
```

In **test** code, `!!` would be preferred as it will fail the test if the data is null. 
```kotlin
@Test 
fun myTest(){
    val whevs = myFixture.someData!!.somefield
    whevs.let{
          val result = doSomething(whevs) 

          assertStuff(result)
    }
}
```

## Context

Generally, `!!`/`requireNotNull`/`checkNotNull` are preferred in tests over `?.`/`if(foo != null)` etc., although even better to have none of those in a test. 

`UnsafeCallOnNullableType` could be giving the wrong advice when applied to test code.

**Comment by cortinico** [comment]
> Detect usage of `?.` **in a test** that could lead to the test doing nothing at all.

I think the root cause is that you're having an `assert*` method inside a conditional environment. `?.let` is just one occurrence of it, but using `if(... != null)` would have the same semantic.

**Comment by BraisGabin** [comment]
As cortinico said this is more related with the conditionals. We know how to measure the _McCabe's Cyclomatic Complexity_ of any method. We could mix this and ensure that any test have a complexity of 1 (or 0, I don't know the base).

The problem that I see here is that this is framework dependant. It would work in a different way in Junit and Spek. And usually we don't want to implement that in our default rules.

I think that we can't include that as a rule but we should provide all the needed pieces to implemet it. Now that you talked about this rule I want it in my projects too.

**Comment by schalkms** [comment]
Yes and you can't be sure that assert* actually asserts the program context.
Furthermore, there are many test frameworks available that work differently.

**Comment by yogurtearl** [comment]
`?.`, `if(... != null)`, `?:` are all probably smells in test code because they could cause a test to pass because of a null somewhere, when it should fail. 

in tests, it is better to fail in the face of null, rather than risk silently passing. 

You could apply this rule to all source code in test source sets to avoid having to deal with specific test frameworks. 
You don't need to analysis `assert` either, `?.`, `if(... != null)`, `?:`  anywhere in `/test` is a smell and should require some explanation/suppression.

**Comment by cortinico** [comment]
> either, ?., if(... != null), ?: anywhere in /test

I think this is too broad to make a rule out of it.

```kotlin
assertEquals(42, myApi.getAnswer()?.getValue())
```

is a legit test code to me.

**Comment by schalkms** [comment]
> I think this is too broad to make a rule out of it.

Agreed. It's probably better to go the [custom rule](https://arturbosch.github.io/detekt/extensions.html#custom-rulesets) way for Michael. You know the special cases of your repo best. The rule is too broad for general use in detekt. Detekt tries to include rules that are relevant for the majority of users. Because of that, we shall close this issue then.
